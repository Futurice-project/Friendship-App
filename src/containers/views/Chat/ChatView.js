import React, { Component } from 'react';
import { connect } from 'react-redux';
import Modal from 'react-native-modal';
import { NavigationActions } from 'react-navigation';
import ReversedFlatList from 'react-native-reversed-flat-list';
import styled from 'styled-components/native';
import {
  KeyboardAvoidingView,
  Platform,
  Text,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import Icon from 'react-native-vector-icons/Ionicons';
import rest from '../../../utils/rest';
import Button from '../../../components/Button';
import PopUpMenu from '../../../components/PopUpMenu';

const mapDispatchToProps = dispatch => ({
  onViewProfile: profileId =>
    dispatch(
      NavigationActions.navigate({
        routeName: 'ProfileUser',
        params: { profileId },
      }),
    ),
  chatRoomMessages: id => {
    dispatch(rest.actions.chatRoomMessages({ id }));
  },
  //update all messages that have been read
  updateReadMessages: messageIdArr => {
    dispatch(
      rest.actions.updateReadMessages(
        {},
        { body: JSON.stringify({ messageIdArr: messageIdArr }) },
      ),
    );
  },
  sendMessage: (id, textMessage, userId) => {
    dispatch(
      rest.actions.sendMessage(
        { id },
        {
          body: JSON.stringify({ textMessage, userId }),
        },
      ),
    );
  },
});

const mapStateToProps = state => ({
  auth: state.auth,
  currentUserId: state.auth.data.decoded ? state.auth.data.decoded.id : null,
  chatRoom: state.chatRoomMessages.data,
  sentMessage: state.sendMessage,
});

class ChatView extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: `${navigation.state.params.userEmoji} ${navigation.state.params
      .username}`,
    headerLeft: (
      <Icon
        style={{ padding: 15, fontSize: 26 }}
        name={'ios-arrow-back'}
        onPress={() => {
          navigation.dispatch(
            NavigationActions.reset({
              index: 0,
              actions: [
                NavigationActions.navigate({
                  routeName: 'InboxView',
                }),
              ],
            }),
          );
        }}
      />
    ),
    headerRight: (
      <PopUpMenu isReportVisible={navigation.state.params.showReport} chat />
    ),
  });

  state = {
    chatroomId: '',
    text: '',
    description: '',
    isOptionsVisible: false,
    isReportVisible: false,
    isReportStatusVisible: false,
    reportStatusText: '',
  };

  componentDidMount = () => {
    this.setState({
      chatroomId: this.props.navigation.state.params.chatroomId,
    });
    this.props.navigation.setParams({ showReport: this.showReport });
    this.props.chatRoomMessages(this.props.navigation.state.params.chatroomId);
    //update all unread messages after 3 seconds to make sure all the chatroom messages have been fetched
    setTimeout(() => this.getUnreadMessagesAndUpdateStatus(), 3000);
  };

  componentWillReceiveProps = () => {
    if (this.props.chatroom) {
      this.props.navigation.setParams({ chatroom: this.props.chatroom });
    }
  };

  getUnreadMessagesAndUpdateStatus = () => {
    //get an array of all the unread messages which have the 'read' field equals to 'false' and user_id not equals to current user id
    let messageArr = this.props.chatRoom.messages
      ? this.props.chatRoom.messages.filter(
          message =>
            message.read === false &&
            message.user_id !== this.props.currentUserId,
        )
      : [];
    //get an array of all the id of unread messages
    let messageIdArr = messageArr.map(message => message.id);
    //call the update function to change the 'read' field into 'true'
    this.props.updateReadMessages(messageIdArr);
  };

  sendMessage = () => {
    //Keyboard.dismiss();
    const chatroomId = this.props.navigation.state.params.chatroomId;
    const textMessage = this.state.text;
    const userId = this.props.currentUserId;

    this.props.sendMessage(chatroomId, textMessage, userId);
    this.setState({ text: '' });
  };

  // Modal functions
  showReport = () => {
    const { isReportVisible } = this.state;
    this.setState({ isReportVisible: !isReportVisible, reportDescription: '' });
  };

  sendReport = () => {
    const creator = this.props.chatRoom.creator;
    const receiver = this.props.chatRoom.receiver;
    const userId =
      this.props.currentUserId === creator.id ? receiver.id : creator.id;
    const description = this.state.reportDescription;
    const reported_by = this.props.currentUserId;
    fetch(`http://localhost:3888/reports`, {
      method: 'post',
      headers: {
        Authorization: this.props.auth.data.token,
      },
      body: JSON.stringify({
        userId,
        description,
        reported_by,
      }),
    })
      .then(() =>
        this.setState({
          isReportStatusVisible: true,
          isReportVisible: false,
          reportStatusText: 'User reported',
        }),
      )
      .catch(() =>
        this.setState({
          isReportStatusVisible: true,
          isReportVisible: false,
          reportStatusText: 'Report failed',
        }),
      );
  };

  keyExtractor = (item, index) => index;

  renderItem = ({ item }) => {
    const textAlign =
      item.user_id === this.props.currentUserId ? 'right' : 'left';
    const messageCardStyle =
      item.user_id === this.props.currentUserId
        ? styles.SendCard
        : styles.ReceiveCard;

    let time = '';
    const msgTime = new Date(item.chat_time);
    if (msgTime) {
      const msgDate = msgTime.getDate();
      const msgMonth = msgTime.getMonth();
      const msgYear = msgTime.getFullYear();
      const now = new Date();
      const diff =
        Math.abs(now.getTime() - msgTime.getTime()) / (1000 * 60 * 60 * 24);

      const days = [
        'Sunday',
        'Monday',
        'Tuesday',
        'Wednesday',
        'Thursday',
        'Friday',
        'Saturday',
      ];
      const months = [
        'Jan',
        'Feb',
        'Mar',
        'Apr',
        'May',
        'Jun',
        'Jul',
        'Aug',
        'Sep',
        'Oct',
        'Nov',
        'Dec',
      ];
      const timeArr = msgTime
        .toTimeString()
        .split(' ')[0]
        .split(':');

      if (now.getFullYear() !== msgYear) {
        //not same year
        time =
          months[msgMonth] +
          ' ' +
          msgDate +
          ' ' +
          msgYear +
          ' - ' +
          timeArr[0] +
          ':' +
          timeArr[1];
      } else if (now.getFullYear() === msgYear && diff > 7) {
        //if not within a week
        time =
          months[msgMonth] +
          ' ' +
          msgDate +
          ' - ' +
          timeArr[0] +
          ':' +
          timeArr[1];
      } else if (now.getFullYear() === msgYear && diff <= 7) {
        //day of week
        time = days[msgTime.getDay()] + ' - ' + timeArr[0] + ':' + timeArr[1];
      } else if (
        now.getFullYear() === msgYear &&
        now.getMonth() === msgMonth &&
        now.getDate() === msgDate
      ) {
        //today
        time = timeArr[0] ? timeArr[0] + ':' + timeArr[1] : '';
      } else if (now.getTime() - msgTime.getTime() < 0) {
        time =
          months[msgMonth] +
          ' ' +
          msgDate +
          ' ' +
          msgYear +
          ' ' +
          msgTime.toTimeString().split(' ')[0];
      }
    }

    return (
      <View style={messageCardStyle}>
        <Text
          style={{
            textAlign,
            fontSize: 10,
            color: '#60686d',
            marginBottom: 10,
          }}
        >
          {time}
        </Text>
        <Text style={{ color: '#4a4a4a', textAlign }}>{item.text_message}</Text>
      </View>
    );
  };

  render() {
    let reportTitle = 'Report ';
    return (
      <KeyboardAvoidingView
        style={{ flex: 1 }}
        behavior="padding"
        keyboardVerticalOffset={Platform.select({
          ios: () => 65,
          android: () => 80,
        })()}
      >
        <ReversedFlatList
          data={this.props.chatRoom.messages || []}
          keyExtractor={this.keyExtractor}
          renderItem={this.renderItem}
          style={[
            { flex: 1 },
            this.state.isReportVisible || this.state.isReportStatusVisible
              ? { backgroundColor: 'rgba(0,0,0,0.5)' }
              : { backgroundColor: 'white' },
          ]}
        />
        <TextInputCard>
          <TextInput
            style={{ flex: 5, paddingLeft: 20 }}
            returnKeyType="send"
            placeholder={'Message '}
            onChangeText={text => this.setState({ text })}
            value={this.state.text}
            onSubmitEditing={() => this.sendMessage()}
          />
          <ChatInputButtonCard>
            <TouchableOpacity onPress={() => this.sendMessage()}>
              <Text
                style={{
                  fontWeight: 'normal',
                  fontSize: 16,
                  color: '#ff8a65',
                  paddingRight: 15,
                }}
              >
                Send
              </Text>
            </TouchableOpacity>
          </ChatInputButtonCard>
        </TextInputCard>
        <Modal
          visible={this.state.isReportVisible}
          transparent
          animationType="slide"
        >
          <View
            style={{
              height: 270,
              borderRadius: 5,
              backgroundColor: '#F1F1F3',
              padding: 20,
              paddingLeft: 10,
            }}
          >
            <View
              style={{
                marginLeft: 10,
                borderBottomWidth: 0.8,
                borderBottomColor: 'gray',
                marginBottom: 20,
                paddingBottom: 20,
              }}
            >
              <Text
                style={{
                  textAlign: 'center',
                  fontSize: 20,
                  fontWeight: 'bold',
                  paddingBottom: 15,
                  borderBottomColor: 'gray',
                  borderBottomWidth: 0.8,
                }}
              >
                Send Report
              </Text>
              <Text style={{ marginTop: 10, fontSize: 16, color: '#60686d' }}>
                Report Description
              </Text>
              <TextInput
                autoCorrect={false}
                autoCapitalize="none"
                title={reportTitle}
                placeholder="Description"
                multiline={true}
                onChangeText={reportDescription =>
                  this.setState({ reportDescription })}
                value={this.state.reportDescription}
                style={styles.reportInput}
              />
              <Text style={{ marginTop: 5, color: '#60686d' }}>
                *Average response time is 2 days{' '}
              </Text>
            </View>
            <View style={{ flexDirection: 'row', justifyContent: 'flex-end' }}>
              <TouchableOpacity
                title="Cancel"
                onPress={this.showReport}
                style={styles.cancelButton}
              >
                <Text
                  style={{
                    color: '#F9F1EF',
                    fontSize: 16,
                    fontWeight: '600',
                  }}
                >
                  Cancel
                </Text>
              </TouchableOpacity>
              <TouchableOpacity
                title="Report"
                onPress={this.sendReport}
                style={styles.reportButton}
              >
                <Text
                  style={{
                    color: '#F9F1EF',
                    fontSize: 16,
                    fontWeight: '600',
                  }}
                >
                  Report
                </Text>
              </TouchableOpacity>
            </View>
          </View>
        </Modal>
        <Modal
          visible={this.state.isReportStatusVisible}
          transparent
          animationType="slide"
        >
          <View
            style={{
              height: 150,
              borderRadius: 10,
              backgroundColor: '#F1F1F3',
              padding: 20,
              paddingLeft: 10,
            }}
          >
            <Text
              style={{ textAlign: 'center', fontSize: 25, marginBottom: 10 }}
            >
              {this.state.reportStatusText}
            </Text>
            <TouchableOpacity
              style={{
                alignSelf: 'stretch',
                padding: 10,
                backgroundColor: 'red',
                borderRadius: 10,
              }}
              onPress={() => this.setState({ isReportStatusVisible: false })}
            >
              <Text style={{ textAlign: 'center' }}>OK</Text>
            </TouchableOpacity>
          </View>
        </Modal>
      </KeyboardAvoidingView>
    );
  }
}

const ChatInputButtonCard = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
`;

const TextInputCard = styled.View`
  height: 44px;
  background-color: #e8e9e8;
  flex-direction: row;
`;

const styles = {
  SendCard: {
    flex: 1,
    padding: 20,
    paddingBottom: 30,
    margin: 10,
    marginRight: 20,
    marginLeft: 40,
  },
  ReceiveCard: {
    flex: 1,
    padding: 20,
    paddingBottom: 30,
    margin: 10,
    marginRight: 40,
    marginLeft: 20,
  },
  reportInput: {
    borderLeftWidth: 0.5,
    borderRightWidth: 0.5,
    borderWidth: 0.5,
    borderTopWidth: 0.5,
    borderRadius: 8,
    height: 40,
    borderBottomColor: 'grey',
    borderTopColor: 'grey',
    borderLeftColor: 'grey',
    borderRightColor: 'grey',
    marginLeft: 2,
    marginTop: 5,
    paddingRight: 5,
    paddingLeft: 5,
    fontSize: 18,
    lineHeight: 23,
  },
  reportButton: {
    backgroundColor: '#00bfff',
    borderRadius: 5,
    borderWidth: 1,
    padding: 13,
    borderColor: '#14B28B',
    marginLeft: 5,
  },
  cancelButton: {
    backgroundColor: '#ed5249',
    borderRadius: 5,
    borderWidth: 1,
    padding: 13,
    borderColor: '#14B28B',
  },
};

export default connect(mapStateToProps, mapDispatchToProps)(ChatView);
