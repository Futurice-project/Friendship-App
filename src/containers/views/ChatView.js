import React, { Component } from 'react';
import { connect } from 'react-redux';
import { NavigationActions } from 'react-navigation';
import {
  TextInput,
  Image,
  FlatList,
  TouchableOpacity,
  KeyboardAvoidingView,
  Platform,
  LayoutAnimation,
  NativeModules,
  Keyboard,
} from 'react-native';
import {
  TextInputCard,
  ChatMessageCard,
  ChatInputButtonCard,
  ChatHeaderImage,
} from '../../components/Layout';
import MessageCard from '../../components/MessageCard';
import PopUpMenu from '../../components/PopUpMenu';

const { UIManager } = NativeModules;

UIManager.setLayoutAnimationEnabledExperimental &&
  UIManager.setLayoutAnimationEnabledExperimental(true);

class ChatView extends Component {
  static navigationOptions = ({ navigation }) => ({
    title: `${navigation.state.params.name}`,
    headerRight: (
      <TouchableOpacity onPress={navigation.state.params.onMenuPopup}>
        <ChatHeaderImage source={require('../../../assets/following.png')} />
      </TouchableOpacity>
    ),
  });

  componentDidMount() {
    this.scrollToBottom();
    this.props.navigation.setParams({ onMenuPopup: this.onMenuPopup });
  }

  state = {
    popUpMenu: false,
    text: '',
    currentUser: 'Thu',
    messages: [
      { user: 'Peter', text: 'Lorem ipsum' },
      { user: 'Thu', text: 'Dolor sit amet' },
      { user: 'Thu', text: 'Consectetur' },
      { user: 'Peter', text: 'adipiscing elit ir' },
      { user: 'Thu', text: 'morbi tristique' },
      { user: 'Peter', text: 'Pellentesque' },
      { user: 'Peter', text: 'convallis cursus' },
      { user: 'Thu', text: 'ullamcorper' },
      { user: 'Peter', text: 'Donec mattis' },
      { user: 'Thu', text: 'adipiscing elit' },
      { user: 'Peter', text: 'mattis nunc' },
      { user: 'Peter', text: 'convallis c3ursus' },
      { user: 'Thu', text: 'ulla5mcorper' },
      { user: 'Peter', text: 'Donec fmattis' },
      { user: 'Thu', text: 'adipisciwng elit' },
      { user: 'Peter', text: 'mattis fnunc' },
      { user: 'Thu', text: 'ulla5mcodrper' },
      { user: 'Peter', text: 'Donec fmaattis' },
      { user: 'Thu', text: 'adipisciwang elit' },
      { user: 'Peter', text: 'mattis fncunc' },
    ],
  };

  onMenuPopup = () => {
    const { popUpMenu } = this.state;
    this.setState({ popUpMenu: !popUpMenu });
    LayoutAnimation.spring();
  };

  onSend = () => {
    const { text, currentUser } = this.state;
    const newMessage = { user: currentUser, text };
    Keyboard.dismiss();
    if (text !== '') {
      this.setState({
        messages: [...this.state.messages, newMessage],
        text: '',
      });
      this.scrollToBottom();
    }
  };

  onViewProfile = () => {
    this.props.onViewProfile();
  };

  onDelete = () => {
    this.props.navigation.navigate('Inbox');
  };

  onBlock = () => {
    this.props.navigation.navigate('Inbox');
  };

  keyExtractor = item => item.text;
  renderItem = ({ item }) => {
    const { text, user } = item;
    return (
      <MessageCard message={text} isSent={user === this.state.currentUser} />
    );
  };

  getItemLayout = (data, index) => ({ length: 50, offset: 50 * index, index });

  scrollToBottom = () => {
    const wait = new Promise(resolve => setTimeout(resolve, 200));
    wait.then(() => {
      this.flatListRef.scrollToIndex({
        animated: true,
        index: this.state.messages.length - 1,
      });
    });
  };

  render() {
    return (
      <KeyboardAvoidingView
        style={{ flex: 1 }}
        behavior="padding"
        keyboardVerticalOffset={Platform.select({
          ios: () => 0,
          android: () => 80,
        })()}
      >
        <ChatMessageCard>
          <PopUpMenu
            popup={this.state.popUpMenu}
            onViewProfile={this.onViewProfile}
            onDelete={this.onDelete}
            onBlock={this.onBlock}
          />
          <FlatList
            data={this.state.messages}
            ref={ref => {
              this.flatListRef = ref;
            }}
            keyExtractor={this.keyExtractor}
            renderItem={this.renderItem}
            style={{ flex: 1 }}
            getItemLayout={this.getItemLayout}
          />
        </ChatMessageCard>
        <TextInputCard>
          <TextInput
            style={{ flex: 5, paddingLeft: 20 }}
            placeholder={'Your message '}
            onChangeText={text => this.setState({ text })}
            value={this.state.text}
            onSubmitEditing={this.onSend}
          />
          <ChatInputButtonCard>
            <TouchableOpacity onPress={() => this.onSend()}>
              <Image
                source={require('../../../assets/following.png')}
                style={{ width: 30, height: 30 }}
              />
            </TouchableOpacity>
          </ChatInputButtonCard>
        </TextInputCard>
      </KeyboardAvoidingView>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  onViewProfile: profileId =>
    dispatch(
      NavigationActions.navigate({
        routeName: 'ProfileUser',
        params: { profileId },
      }),
    ),
});

export default connect(null, mapDispatchToProps)(ChatView);
