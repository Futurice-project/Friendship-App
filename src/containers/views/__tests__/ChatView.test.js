import React from 'react';
import 'react-native';
import { shallow } from 'enzyme';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import ChatView from '../Chat/ChatView';

describe('ChatView', () => {
  const mockStore = configureMockStore([thunk]);

  const messages = [
    {
      id: 23,
      text_messages: 'Hello',
      chat_time: '05-04-2018',
      user_id: 3,
      chatroom_id: 4,
      read: false,
    },
    {
      id: 24,
      text_messages: 'How are you',
      chat_time: '05-05-2018',
      user_id: 4,
      chatroom_id: 4,
      read: true,
    },
  ];

  const initialState = {
    auth: {
      data: {
        decoded: {
          id: 1,
        },
      },
    },
    chatRoomMessages: {
      data: {
        id: 4,
        creator: 'test',
        receiver: 'test',
        message: messages,
      },
    },
  };
  test('render correctly', () => {
    const store = mockStore(initialState);
    const wrapper = shallow(<ChatView store={store} />);

    expect(wrapper).toMatchSnapshot();
  });
});
