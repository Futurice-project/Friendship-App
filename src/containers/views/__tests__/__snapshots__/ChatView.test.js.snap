// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatView render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ChatView)
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "auth": Object {
        "data": Object {
          "decoded": Object {
            "id": 1,
          },
        },
      },
      "chatRoom": Object {
        "creator": "test",
        "id": 4,
        "message": Array [
          Object {
            "chat_time": "05-04-2018",
            "chatroom_id": 4,
            "id": 23,
            "read": false,
            "text_messages": "Hello",
            "user_id": 3,
          },
          Object {
            "chat_time": "05-05-2018",
            "chatroom_id": 4,
            "id": 24,
            "read": true,
            "text_messages": "How are you",
            "user_id": 4,
          },
        ],
        "receiver": "test",
      },
      "chatRoomMessages": [Function],
      "currentUserId": 1,
      "onViewProfile": [Function],
      "sendMessage": [Function],
      "sentMessage": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
      "updateReadMessages": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "auth": Object {
          "data": Object {
            "decoded": Object {
              "id": 1,
            },
          },
        },
        "chatRoom": Object {
          "creator": "test",
          "id": 4,
          "message": Array [
            Object {
              "chat_time": "05-04-2018",
              "chatroom_id": 4,
              "id": 23,
              "read": false,
              "text_messages": "Hello",
              "user_id": 3,
            },
            Object {
              "chat_time": "05-05-2018",
              "chatroom_id": 4,
              "id": 24,
              "read": true,
              "text_messages": "How are you",
              "user_id": 4,
            },
          ],
          "receiver": "test",
        },
        "chatRoomMessages": [Function],
        "currentUserId": 1,
        "onViewProfile": [Function],
        "sendMessage": [Function],
        "sentMessage": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
        "updateReadMessages": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
